operators/pkg/tenant-controller/tenant_controller.go

```diff
@@ -278,6 +267,30 @@ func (r *TenantReconciler) handleDeletion(ctx context.Context, tnName string) er
 	return retErr
 }
 
+// checkValidWorkspaces check validity of workspaces in tenant and updates resource quota in the tenant status.
+func (r *TenantReconciler) checkValidWorkspaces(ctx context.Context, tn *crownlabsv1alpha2.Tenant) ([]crownlabsv1alpha2.TenantWorkspaceEntry, error) {
+	tenantExistingWorkspaces := []crownlabsv1alpha2.TenantWorkspaceEntry{}
+	workspaces := []crownlabsv1alpha1.Workspace{}
+	tn.Status.FailingWorkspaces = []string{}
+	var err error
+	// check every workspace of a tenant
+	for _, tnWs := range tn.Spec.Workspaces {
+		wsLookupKey := types.NamespacedName{Name: tnWs.Name}
+		var ws crownlabsv1alpha1.Workspace
+		if err = r.Get(ctx, wsLookupKey, &ws); err != nil {
+			// if there was a problem, add the workspace to the status of the tenant
+			klog.Errorf("Error when checking if workspace %s exists in tenant %s -> %s", tnWs.Name, tn.Name, err)
+			tn.Status.FailingWorkspaces = append(tn.Status.FailingWorkspaces, tnWs.Name)
+			tnOpinternalErrors.WithLabelValues("tenant", "workspace-not-exist").Inc()
+		} else {
+			tenantExistingWorkspaces = append(tenantExistingWorkspaces, tnWs)
+			workspaces = append(workspaces, ws)
+		}
+	}
+	tn.Status.Quotas = forge.TenantResourceList(workspaces, tn.Spec.Quotas)
```

Comment:
Please, avoid side effects, which can lead to unexpected behavior. If this function is named `checkValidWorkspaces` (which is perfectly fine), it does not deal with resource computations. That should be moved outside of this function.